llama_index/agent/openai_agent.py

```diff
@@ -57,267 +61,231 @@ def call_function(
+
+    def reset(self) -> None:
+        self.memory.reset()
+
+    def prepare_message(self, message: str) -> Tuple[list[BaseTool], list[dict]]:
+        """Prepare tools and functions for the message."""
+        self.tools = self.get_tools_callback(message)
+        self.functions = [tool.metadata.to_openai_function() for tool in self.tools]
+        return self.tools, self.functions
+
+    def get_all_messages(self) -> list[ChatMessage]:
+        return self.prefix_messages + self.memory.get()
 
+    def get_latest_function_call(self) -> Optional[dict]:
+        return self.memory.get_all()[-1].additional_kwargs.get("function_call", None)
+
+
+class BaseOpenAIAgent(BaseAgent):
     def __init__(
         self,
         llm: OpenAI,
         memory: BaseMemory,
-        prefix_messages: List[ChatMessage],
-        verbose: bool = False,
-        max_function_calls: int = DEFAULT_MAX_FUNCTION_CALLS,
-        callback_manager: Optional[CallbackManager] = None,
-    ) -> None:
+        prefix_messages: list[ChatMessage],
+        verbose: bool,
+        max_function_calls: int,
+        callback_manager: Optional[CallbackManager],
+    ):
         self._llm = llm
-        self._memory = memory
-        self._prefix_messages = prefix_messages
         self._verbose = verbose
         self._max_function_calls = max_function_calls
         self.callback_manager = callback_manager or CallbackManager([])
+        self.session = ChatSession(memory, prefix_messages, self._get_tools)
+        self.sources: list[ToolOutput] = []
 
     @property
-    def chat_history(self) -> List[ChatMessage]:
-        return self._memory.get_all()
+    def chat_history(self) -> list[ChatMessage]:
+        return self.session.chat_history
+
+    @property
+    def all_messages(self) -> list[ChatMessage]:
+        return self.session.get_all_messages()
 
     def reset(self) -> None:
-        self._memory.reset()
+        self.session.reset()
 
     @abstractmethod
-    def _get_tools(self, message: str) -> List[BaseTool]:
+    def _get_tools(self, message: str) -> list[BaseTool]:
         """Get tools."""
-
-    def _get_latest_function_call(
-        self, chat_history: List[ChatMessage]
-    ) -> Optional[dict]:
-        """Get latest function call from chat history."""
-        return chat_history[-1].additional_kwargs.get("function_call", None)
-
-    def _init_chat(self, message: str) -> Tuple[List[BaseTool], List[dict]]:
-        """Add user message to chat history and get tools and functions."""
-        self._memory.put(ChatMessage(content=message, role=MessageRole.USER))
-        tools = self._get_tools(message)
-        functions = [tool.metadata.to_openai_function() for tool in tools]
-        return tools, functions
-
-    def chat(
-        self, message: str, chat_history: Optional[List[ChatMessage]] = None
-    ) -> AgentChatResponse:
+        pass
+
+    def _should_continue(
+        self, function_call: dict | None, n_function_calls: int
```

Comment:
Looks like this syntax is breaking the docs? I think the preferred/working syntax would be `Optional[dict]`

Reply:
D'oh, just saw. My bad, been using 3.10+ type annotations by force of habit. Will push fix.

Reply:
Thanks!